src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs

```diff
@@ -0,0 +1,196 @@
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(start =>
+            {
+                var state = new CompilationState(
+                    Flatten(
+                        start.Compilation.GetTypeByMetadataName(Types.Package.FullName)?.GetMembers(Types.Package.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.AsyncPackage.FullName)?.GetMembers(Types.AsyncPackage.GetServiceAsync),
+                        start.Compilation.GetTypeByMetadataName(Types.AsyncPackage.FullName)?.GetMembers(Types.Package.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.ServiceProvider.FullName).GetMembers(Types.ServiceProvider.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.IServiceProvider.FullName).GetMembers(Types.IServiceProvider.GetService),
+                        start.Compilation.GetTypeByMetadataName(Types.IAsyncServiceProvider.FullName).GetMembers(Types.IAsyncServiceProvider.GetServiceAsync)),
+                    Flatten(
+                        start.Compilation.GetTypeByMetadataName(Types.Assumes.FullName).GetMembers(Types.Assumes.Present)));
+                if (state.ShouldAnalyze)
+                {
+                    start.RegisterSyntaxNodeAction(state.AnalyzeInvocationExpression, SyntaxKind.InvocationExpression);
+                }
+            });
+        }
+
+        private static ImmutableArray<ISymbol> Flatten(params ImmutableArray<ISymbol>?[] symbols) => symbols.Where(m => m.HasValue).SelectMany(m => m.Value).ToImmutableArray();
+
+        private class CompilationState
+        {
+            /// <summary>
+            /// GetService or GetServiceAsync methods.
+            /// </summary>
+            private readonly ImmutableArray<ISymbol> getServiceMethods;
+
+            /// <summary>
+            /// Methods that throw when their first argument is null.
+            /// </summary>
+            private readonly ImmutableArray<ISymbol> nullThrowingMethods;
+
+            internal CompilationState(ImmutableArray<ISymbol> getServiceMethods, ImmutableArray<ISymbol> nullThrowingMethods)
+            {
+                this.getServiceMethods = getServiceMethods;
+                this.nullThrowingMethods = nullThrowingMethods;
+            }
+
+            internal bool ShouldAnalyze => !this.getServiceMethods.IsEmpty;
+
+            internal void AnalyzeInvocationExpression(SyntaxNodeAnalysisContext context)
+            {
+                var invocationExpression = (InvocationExpressionSyntax)context.Node;
+                var invokedMethod = context.SemanticModel.GetSymbolInfo(invocationExpression.Expression, context.CancellationToken).Symbol;
+                if (invokedMethod != null && this.getServiceMethods.Contains(invokedMethod))
+                {
```

Comment:
tagging @mavasani @jinujoseph 

this basically register triggering part, and report diagnostics in bigger scope (non local diagnostics).

IDE currently won't handle this good except the most common one where it reports local diagnostics..

should we invest in this area more?

Reply:
@heejaechang This is theoretically a nullness dataflow problem, and ideally should be implemented as a dataflow analyzer. Until we have full support for dataflow analysis, I feel this analyzer should be implemented as either a _code block analyzer_ or an _operation block analyzer_, as the analysis and diagnostic scope is actually much broader that the registered callback unit.

1. A [code block analyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,92) that [registers a nested syntax node action](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,771) to perform analysis and stores mutable state and uses [code block end action](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,750) to report diagnostics based on the this mutable state.
2. An [operation block analyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,137) that [registers a nested operation action](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,922) with [OperationKind.Invocation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/OperationKind.cs,57) to perform analysis and stores mutable state and uses [operation block end action](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,901) to report diagnostics based on this mutable state.

Reply:
Can you elaborate on this more? What about this method is "non local diagnostics" and what does this mean and how does the IDE not handle it well?

Reply:
local diagnostics means, span you analyze and span you report diagnostics are same. since analysis is done in this directions (analysis span -> diagnostic reported span), and fix is done in opposite directions (diagnostic reported span -> analysis span)

local diagnostics are easy to handle.

on another hand, non local diagnostics, is tricky, since only way to see whether the non local diagnostic is still valid is reanalyzing whole project since we don't know which span we need to re-analyze to generate the diagnostics in hand.

saving who created this diagnostics from which span doesn't do much since that doesn't means other span can't generate this diagnostics now.

anyway, due to how expansive it is to maintain non local diagnostics, we normally won't do any non-local diagnostic calculation/reporting/fixing if full solution analysis is off. even when it is on, we don't do it right away.